admin people pagination frontend

alxnddr opened the pull request:
### Description

This PR adds pagination on the people and the group members page. Search is added only to the people page.

#### People page 
Url: http://localhost:3000/admin/people

<img width="1437" alt="Screen Shot 2021-04-28 at 01 35 39" src="https://user-images.githubusercontent.com/14301985/116321356-8b731080-a7c2-11eb-94aa-37bf1e346feb.png">

#### Group members mage
Url: http://localhost:3000/admin/people/groups/1

<img width="1440" alt="Screen Shot 2021-04-28 at 01 36 04" src="https://user-images.githubusercontent.com/14301985/116321362-8f069780-a7c2-11eb-89fa-53aab053a26b.png">

alxnddr committed:
admin people pagination frontend

kulyk commented on frontend/src/metabase/admin/people/components/SearchInput.jsx:
```diff
@@ -0,0 +1,105 @@
+  value: PropTypes.string,
+  autoFocus: PropTypes.bool,
+  className: PropTypes.string,
+  clearButton: PropTypes.bool,
```
`clearButton` prop makes me think it can also be a React node. Should we maybe prefix it like `hasClearButton`?

In general, I think it's a good practice to prefix boolean variables with verbs like "has" or "is"; it makes their type clear. Maybe it's worth adding to our style guide? Unless you agree

kulyk commented on frontend/src/metabase/admin/people/components/UserGroupSelect.jsx:
```diff
@@ -1,58 +1,70 @@
+  render() {
+    const { groups, userMemberships, isCurrentUser } = this.props;
+
+    if (!groups || groups.length === 0 || !userMemberships) {
       return <LoadingSpinner />;
     }
```
Is it possible there are no groups and `groups.length === 0` is part of a _loaded_ behavior?
Entity loaders provide loading props like `loading` and `loaded`, should we maybe rely them?

kulyk commented on frontend/src/metabase/admin/people/const.js:
```diff
@@ -0,0 +1,4 @@
+export const USER_STATUS = {
+  active: "active",
+  deactivated: "deactivated",
+};
```
Other files with constants are called `constants`. Should we use the same name here, maybe?

kulyk commented on frontend/src/metabase/admin/people/selectors.js:
```diff
@@ -29,6 +29,12 @@ export const getUsersWithMemberships = createSelector(
     })),
 );
 
+export const getUserMemberships = createSelector(
+  [(_, props) => props.userId, getMemberships],
+  (userId, memberships) =>
+    memberships && _.values(memberships).filter(m => m.user_id === userId),
+);
+
```
From our [style guide wiki](https://github.com/metabase/metabase/wiki/Frontend%3A-Style-Guide#additional-javascript-conventions), I see we prefer using underscore functions only for things that are not implemented in the standard library.

> Prefer native Array methods over underscore's. We polyfill all ES6 features. Use Underscore for things that aren't implemented natively.

Should we maybe use `Object.values(memberships)` here?

kulyk commented on frontend/test/metabase/components/PaginationControls.unit.spec.js:
```diff
@@ -0,0 +1,63 @@
+  };
+};
+
+describe("PaginationControls", () => {
+  it("should disable pagination buttons when no callbacks provided", () => {
```
That's awesome we have more unit tests, thank you!

kulyk reviewed (approved):
It looks and works great! Have a few minor comments, but in general it's good to go I guess

kulyk commented:
Noticed a small UX issue that we can probably improve

When I go to the last pagination page, there is no easy way to go back. The router is unaware of pagination, so using the browser back button navigates me to the previous page. I think the easiest way to avoid that is not to display the "Next" button if the next page is empty (we have the `total` value, right?). But ideally, we should allow going back with a browser button, as I think.

**Demo**

The demo uses the default H2 database with default users, so it should be easy to reproduce

![2021-04-29 13 59 51](https://user-images.githubusercontent.com/17258145/116541437-228bb580-a8f4-11eb-831f-b37d4351eb55.gif)

alxnddr commented:
@kulyk it should be working correctly because it relies on the `total` value coming from the back end. The total number should also be visible in the page footer but on your recording, it is not. So it looks like you might forget to restart the back end on your local machine when you switched the branch, please check

kulyk commented:
@alxnddr you were right, my bad. Restarted both backend and frontend and now everything works correctly
Good job! Sorry for the false alarm

alxnddr committed:
admin people pagination and search specs

alxnddr committed:
address review comments

alxnddr committed:
fix rebase

alxnddr commented:
@kulyk thanks for the review! Good catch with not resetting the pagination state when search query or active/deactivated filter change. I addressed your comments, please take a look

kulyk commented:
@alxnddr I'm glad it was helpful! Everything looks good to me. Great work!

nemanjaglumac commented on frontend/test/metabase/scenarios/admin/people/people.cy.spec.js:
```diff
@@ -25,13 +30,9 @@ describe("scenarios > admin > people", () => {
-        .as("result-rows")
-        // Bobby Tables, No Collection Tableton, No Data Tableton, None Tableton, Robert Tableton
-        .should("have.length", TOTAL_USERS);
+      cy.findByText("8 people found");
```
Why hard code this value?
If we change the default snapshot in the future (very likely), this will fail.

alxnddr replied:
Fixed all of these, thanks!

nemanjaglumac commented on frontend/test/metabase/scenarios/admin/people/people.cy.spec.js:
```diff
@@ -187,6 +183,78 @@ describe("scenarios > admin > people", () => {
+        // Page 1
+        cy.findByText("1 - 15");
+        assertTableRowsCount(15);
+        cy.findByTestId("previous-page-btn").should("be.disabled");
```
You can alias this and then later use it in the test:
```suggestion
        cy.findByTestId("previous-page-btn")
          .as("prev")
          .should("be.disabled");
```

and to use it later:
`cy.get("@prev").click();`

alxnddr replied:
If you don't have objections I'd prefer to leave these as `previous-page-btn` because `prev` is a bit less descriptive and can mean the previous page

nemanjaglumac replied:
Ofc, do whatever makes more sense to you.
This comment had more to do with the aliasing than what exact string will you use. In a small test like this it doesn't make much difference, but for longer tests it might be easier later to do `cy.get("@alias");`

alxnddr commented on frontend/src/metabase/admin/people/components/SearchInput.jsx:
```diff
@@ -0,0 +1,105 @@
+  hasClearButton: false,
+};
+
+const Input = styled.input`
```
@kdoh I was not sure if we have a defined Input in our design system. There are multiple inputs at the moment including:
- `frontend/src/metabase/components/Input.jsx` — this one has different styles and it uses `Button` css classes inside
- `frontend/src/metabase/components/InputBlurChange.jsx`
- `frontend/src/metabase/admin/datamodel/components/FormInput.jsx` — form input which is currently expects `field` form object
- The one is used inside `AccordionList`
- Some custom `<input type="text" ... />`

I'd love to reuse an existing one but I did not found one. I think we need to unify our UI, improve our UI kit and discourage creating custom components in place. I haven't done it here because premature unification or limiting abstraction without knowing all requirements, use cases for the component is worse than creating a new isolated input in place. What do you think?

daltojohnso reviewed (changes requested):
looks good overall. i know it's sometimes hard to do, but going forward can you try to split up these big ones? not necessarily into separate PRs, but separate commits can be nice to see what code is just getting moved around

alxnddr committed:
address review comments

alxnddr commented on frontend/src/metabase/admin/people/hooks/use-people-query.js:
```diff
@@ -0,0 +1,73 @@
+
+  const [searchText, setSearchText] = useState("");
+
+  useEffect(() => {
```
Yes, adding something like https://github.com/streamich/react-use might be a good option. I did not have a strong need here but after we start using more functional components it will be handy

daltojohnso reviewed (approved):
nice work!

nemanjaglumac reviewed (approved):
I didn't go too deep in the second review of tests, but you did address the changes requested and CI is passing so... YAY! 🎉

howonlee reviewed (approved):
it worked when i futzed around with it a fair bit, no comments that weren't already made
